Replace text angular editor with trix editor in About Us and Shopfront message fields

cillian opened the pull request:
#### What? Why?

- For https://github.com/openfoodfoundation/openfoodnetwork/issues/9481 and https://github.com/openfoodfoundation/openfoodnetwork/issues/9490 

The trix editor was added in https://github.com/openfoodfoundation/openfoodnetwork/pull/11140, this PR follows on from that and replaces the remaining instances of the text-angular editor.

#### What should we test?

- Test the Trix editor is working for the following fields...
  * Enterprise Settings > About > About us 
  * Enterprise Settings > Shop preferences > Shopfront message
  * Enterprise Settings > Shop preferences > Shopfront closed message
  * Enterprise Group Settings > About > About
 
#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

**Before**

![about-us-before](https://github.com/user-attachments/assets/afe8c761-77a0-4be4-ad3f-464815bd9b03)

![shop-messages-before](https://github.com/user-attachments/assets/57ccf6a1-7f41-4ff0-b903-cf7c0ce253fb)

![groups-about-us-before](https://github.com/user-attachments/assets/3b79588a-e95b-4da3-bab1-c601d6476471)

**After** 

![about-us-after](https://github.com/user-attachments/assets/dd7fc3bf-4999-4826-8f5f-7a2414174ddd)

![shop-messages-after](https://github.com/user-attachments/assets/38ec97cf-c487-47ef-9718-1403b2fb9ee8)

![groups-about-us-after](https://github.com/user-attachments/assets/d81940c4-9601-45b4-acb4-bd32146141ca)

Note, in the above screenshots the trix fields are on new lines instead of being alongside the field label like they usually are. This is because there was a horizontal scrollbar if it was beside the field label, see:

![shop-messages-after-with-horizontal-scollbar](https://github.com/user-attachments/assets/87e3d410-83c8-461b-b2c5-968449b46877)

I also added some wrapping to CSS to the trix editor button groups so they will automatically wrap and stop the horizontal scrollbar on smaller screens, although another solution might be to remove the whitespace to the right, see:

![shop-messages-wrapping-after](https://github.com/user-attachments/assets/0957dd90-7a20-46ef-b3d0-b1135e3f23a9)

cillian commented on app/models/enterprise.rb:
```diff
@@ -251,12 +251,21 @@ def self.distinct_count
   # Remove any unsupported HTML.
   def long_description
-    HtmlSanitizer.sanitize(super)
+    HtmlSanitizer.sanitize_and_enforce_link_target_blank(super)
```
I introduced a `sanitize_and_enforce_link_target_blank` method because I wasn't sure if  `target="_blank"`  is needed _all_ the time, or just _some_ of the time.

One other thing I wasn't sure about was whether to use [HtmlSanitizer](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/services/html_sanitizer.rb) or [TrixSanitizer](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/services/trix_sanitizer.rb), they both seem similar but they are not the exact same.

rioug replied:
From memory `HtmlSanitizer` was introduced specifically for `Enterprise` model so this is the correct place to add `sanitize_and_enforce_link_target_blank`.

dacook replied:
Sounds good to me. I'm guessing that the previous solution did the same thing? (enforce link `target="blank"` on _all_ links). 
I'm guessing it was because the links would normally point to an external website. It raises the question what if people use internal links to the current website. But that should be answered in a separate issue, if needed.

cillian replied:
Yes, previously the text angular editor would add a `target="_blank"` attribute every time you inserted a link, see https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/assets/javascripts/admin/utils/directives/textangular_links_target_blank.js.coffee

rioug reviewed (approved):
yeah less angular :broom:  ! Thanks @cillian :pray:

dacook commented on app/views/admin/enterprise_groups/_form_about.html.haml:
```diff
@@ -1,6 +1,5 @@
-    %text-angular{'id' => 'enterprise_group_long_description', 'name' => 'enterprise_group[long_description]', 'class' => 'text-angular', "textangular-links-target-blank" => true,
-      'ta-toolbar' => "[['h1','h2','h3','h4','p'],['bold','italics','underline','clear'],['insertLink']]"}
-      != @enterprise_group[:long_description]
+    = f.hidden_field :long_description, id: "enterprise_group_long_description", value: @enterprise_group.long_description
```
If the subject of the form `f` is `@enterprise_group`, then the ID and Value should be assigned automatically. Does it work if they are omitted?

```suggestion
    = f.hidden_field :long_description, id: "enterprise_group_long_description"
```

(Same for others below)

dacook commented on config/locales/en.yml:
```diff
@@ -3725,6 +3723,33 @@ See the %{link} to find out more about %{sitename}'s features and to start using
         There was a problem setting up your card in our payments gateway.
         Please refresh the page and try again, if it fails a second time,
         please contact us for support.
+    trix:
+      attachFiles: "Attach Files"
+      bold: "Bold"
+      bullets: "Bullets"
+      byte: "Byte"
+      bytes: "Bytes"
+      captionPlaceholder: "Add a caption…"
+      code: "Code"
+      heading1: "Heading"
+      indent: "Increase Level"
+      italic: "Italic"
+      link: "Link"
+      numbers: "Numbers"
+      outdent: "Decrease Level"
+      quote: "Quote"
+      redo: "Redo"
+      remove: "Remove"
+      strike: "Strikethrough"
+      undo: "Undo"
+      unlink: "Unlink"
+      url: "URL"
+      urlPlaceholder: "Enter a URL…"
+      GB: "GB"
+      KB: "KB"
+      MB: "MB"
+      PB: "PB"
+      TB: "TB"
```
Thanks for adding Trix translation keys. 

These two old labels seem equivalent to new labels (I haven't checked though)
* "Insert / edit link" -> "Link"
* "Please enter a URL to insert" -> "Enter a URL…"

If so, I think it would help to use the old labels. That way, it will probably be easy for translators to copy the existing labels when adding translations for Trix. Does that make sense?

cillian replied:
Good idea, that's updated in https://github.com/openfoodfoundation/openfoodnetwork/pull/12734/commits/2d6ffc0ca16459f0d4ee3abdaf82189ddd0a8656 to use the old "Please enter a URL to insert" label.

The "Link" translation is used in two places at once so the "Insert / edit link" label is a bit more awkward, see...

![link-translation](https://github.com/user-attachments/assets/4e9d22f2-81ee-4016-88a7-71c8ec7fcbcc)

dacook reviewed (commented):
Great work, it looks much cleaner than before!
I like how you adapted to fit the extra Trix buttons in, and it means there's more space for the text area too. 

I have a couple of comments that might be worth trying. Do they make sense? 
If not, we can proceed as is.

cillian commented on app/webpacker/controllers/trixeditor_controller.js:
```diff
@@ -42,4 +44,57 @@ export default class extends Controller {
+      for (const [key, translation] of Object.entries(I18n.t("js.trix"))) {
+
+        // Set all translations (only works in Firefox for some reason)
+        Trix.config.lang[key] = translation;
+
+        // Set toolbar translations (Chrome)
+        this.#setTranslation(
+          `[data-trix-action="${this.#translationKeyToAttributeMappings(key)}"]`,
+          "title",
+          translation
+        );
+        this.#setTranslation(
+          `[data-trix-attribute="${this.#translationKeyToAttributeMappings(key)}"]`,
+          "title",
+          translation
+        );
+      }
+
+      // Set translations for link dialog (Chrome)
+      this.#setTranslation(`[data-trix-dialog="href"] input`,
+                                  "aria-label", I18n.t("js.trix.url"));
+      this.#setTranslation(`[data-trix-dialog="href"] input`,
+                                  "placeholder", I18n.t("js.trix.urlPlaceholder"));
+      this.#setTranslation('.trix-dialog--link input[data-trix-method="setAttribute"]',
+                                  "value", I18n.t("js.trix.link"));
+      this.#setTranslation('.trix-dialog--link input[data-trix-method="removeAttribute"]',
+                                  "value", I18n.t("js.trix.unlink"));
```
` Trix.config.lang[key] = translation;` only seems to work in Firefox for some reason. The translations have to be set more manually for Chrome.

dacook replied:
That's interesting, and unfortunate because you have map quite a lot data.

According to this wiki, it suggests overwriting the `lang` config in one command.
> `Trix.config.lang = translation`
https://github.com/basecamp/trix/wiki/I18n

But you tried that and it didn't work, so it would be worth raising an issue.
I had a look for issues and could only find one. 
The final comment shows another method using `Object.assign`. I don't know if that would make a different, but perhaps it's worth trying.
https://github.com/basecamp/trix/issues/874#issuecomment-866693637

cillian replied:
Good find, fixed by https://github.com/openfoodfoundation/openfoodnetwork/pull/12734/commits/63c62cae08ec22fe234ae0e1b8c3dcde4ac30bec

dacook commented on config/locales/en_AU.yml:
```diff
@@ -2957,6 +2957,24 @@ en_AU:
           There was a problem setting up your card in our payments gateway.
           Please refresh the page and try again, if it fails a second time,
           please contact us for support.
+    trix:
+      bold: "Bold"
+      bullets: "Bullets"
+      code: "Code"
+      heading1: "Heading"
+      hr: "Horizontal rule"
+      indent: "Increase Level"
+      italic: "Italic"
+      link: "Link"
+      numbers: "Numbers"
+      outdent: "Decrease Level"
+      quote: "Quote"
+      redo: "Redo"
+      strike: "Strikethrough"
+      undo: "Undo"
+      unlink: "Unlink"
+      url: "URL"
+      urlPlaceholder: "Please enter a URL to insert"
```
Actually, these will already have the default loaded from `en.yml`

In general , we need to avoid changing any locale files other than the default, as described here:
https://github.com/openfoodfoundation/openfoodnetwork/wiki/Internationalisation-(i18n)#development

I have to admit, I don't know if this commit will be a problem or not, but I know it's not necessary so perhaps it's safest to remove it.

dacook commented on config/locales/ar.yml:
```diff
@@ -3325,6 +3325,8 @@ ar:
           كانت هناك مشكلة في إعداد بطاقتك في بوابة المدفوعات الخاصة بنا.
           يرجى تحديث الصفحة والمحاولة مرة أخرى ، إذا فشلت مرة أخرى ،
           يرجى الاتصال بنا للحصول على الدعم.
+    trix:
+      urlPlaceholder: "الرجاء إدخال عنوان URL لإدراجه"
```
Sorry I just meant to update it in `en.yml`. In general we should avoid changing other locale files directly. I'm unsure if adding new keys is a problem though.

dacook replied:
@mkllnk can you help with a query here? 
For convenience, Cillian has copied an existing translation to a new key (the existing translation was for angular editor, and there is an equivalent in the new trix editor). Do you know if it's ok, and if it's helpful to add new keys to the locale files?
Or should we remove this before merging?

mkllnk replied:
We should remove this addition. Transifex will handle it. It recognises existing translations.

mkllnk commented on config/locales/ar.yml:
```diff
@@ -1051,8 +1051,6 @@ ar:
               رسالة اختيارية للترحيب بالعملاء وشرح كيفية التسوق معك. إذا تم إدخال
               النص هنا ، فسيتم عرضه في علامة تبويب الصفحة الرئيسية عندما يصل العملاء
               إلى واجهة المتجر الخاصة بك.
-          shopfront_message_link_tooltip: "إدراج / تحرير الرابط"
-          shopfront_message_link_prompt: "الرجاء إدخال عنوان URL لإدراجه"
```
Please remove all changes to locales other than en.yml.

cillian replied:
:+1: Fixed by https://github.com/openfoodfoundation/openfoodnetwork/pull/12734/commits/dd5175558e7123052e900ea00b93d29df8a2cc4f

mkllnk reviewed (changes requested):
Great work! Thank you.

We just need to revert he changes to the other locales. Otherwise this is great!

cillian commented:
Thanks. I have reverted those changes to non `en` locale files now and simplified how the Trix translations are set, let me know if you want me to squash the commits.

dacook reviewed (approved):
Great, all ready to test!

drummer83 commented:
@cillian Github is saying there are changes requested. Could you please have a look? I assume it's been addressed already?

mkllnk reviewed (approved):
> Github is saying there are changes requested.

Github was waiting for my review. But the other approvals should have been enough. Anyway, you get may approval, too. :smile:

drummer83 commented:
Hi @cillian
Sorry, there's a conflict here. Could you please have a look?
Thanks!

drummer83 commented:
For future testing I have documented the "before" state already.

- Already existing Trix editor for White Label Custom Tab Content:
![image](https://github.com/user-attachments/assets/930066c1-4ad6-4586-9498-41f65bdf2591)

- Same on narrow screens (scroll bar and still truncated buttons):
![image](https://github.com/user-attachments/assets/59bd5352-b928-4815-9b83-5982b2571818)

- Old editor for Home and Shopfront Closed message with test content using all available options:
![image](https://github.com/user-attachments/assets/072ff0f7-1374-464c-89c6-b28738972591)

- What it looks like in the shopfront:
![image](https://github.com/user-attachments/assets/e7b59fc5-72b2-43f3-a083-80f6763e705b)

cillian commented:
Thanks @drummer83 I have fixed that conflict now. 

For reference, it was this...

```
  # Remove any unsupported HTML.
<<<<<<< HEAD
  def long_description
    HtmlSanitizer.sanitize_and_enforce_link_target_blank(super)
  end

  # Remove any unsupported HTML.
=======
>>>>>>> master
  def long_description=(html)
    super(HtmlSanitizer.sanitize_and_enforce_link_target_blank(html))
  end
```

To fix it I removed the `#long_description` method like what was done in https://github.com/openfoodfoundation/openfoodnetwork/commit/d061fe8ad926e60294b68b9c5dfdfc4d37a20c5b because the enforcing of `target="_blank"` will be added to links when assigning the `long_description=` attribute. Before this PR all links already had `target="_blank"` on them because that was the default behaviour of the old angular text editor.

drummer83 commented:
Hi @cillian,
Thanks for working on this one! Here are my test results after staging.

## After staging the PR

- Shopfront display remains unchanged.
- Back office editor is updated and styling is adapted to newly available options (One header only, no underline, ...).
![grafik](https://github.com/user-attachments/assets/bac78a59-1a77-416c-85f7-b4a46f2d33d4)

## After staging the PR and updating the content
- Shopfront display is adapted to use newly available options.
![grafik](https://github.com/user-attachments/assets/4766a8b5-2161-4bd9-a10c-5dd21d189d74)
- Behavior on narrow screens is much better now.
![grafik](https://github.com/user-attachments/assets/f75e6be2-bd82-4d2b-9868-71b025448c60)

I find it surprising to see, that it's enough to edit any of the content to have the new styles applied to all of the occurences. It's no problem, maybe even an advantage looking at consistency. I assume this happens because when clicking Update, all of the tabs are saved and not just the tab I'm looking at. 

## Summary
- Styling changes in the shop front due to different styling options in the new editor.
- This is not critical, because all information is still available.
- Small screens are much better supported.
- No problems found.

We can merge this one, but the branch is out of date and I can't rebase due to a conflict. I will move this to Ready To Go but add the feedback needed label to have someone else have a look at it. Maybe @filipefurtad0 or @sigmundpetersen could have a look? 🙏 

Thanks again Cillian! 🚀

sigmundpetersen commented:
Merging is blocked because reapproval is needed after Cillian resolved the last conflict.
Maybe @mkllnk or @rioug could have a quick look when coming back to work and merge if everything looks Ok 👍
